persist/storer.go

```
@@ -3,168 +3,222 @@ package persist
 	// Create integrations Storer directory
-	if os.MkdirAll(baseDir, dirFilePerm) != nil {
+	if os.MkdirAll(dir, dirFilePerm) != nil {
 		baseDir = os.TempDir()
 	}
-	return filepath.Join(baseDir, fmt.Sprint(integrationName, ".json"))
+	return baseDir
 }
 
-// NewInMemoryStore will create and initialize an in-memory Storer.
+// NewInMemoryStore will create and initialize an in-memory Storer (not persistent).
 func NewInMemoryStore() Storer {
 	return &inMemoryStore{
-		Data:       make(map[string]interface{}),
+		cachedData: make(map[string]jsonEntry),
+		Data:       make(map[string][]byte),
 		Timestamps: make(map[string]int64),
 	}
 }
 
-// NewFileStore will create and initialize a disk-backed Storer.
-func NewFileStore(storePath string, l log.Logger) (Storer, error) {
-	store := &fileStore{
-		inMemoryStore: *NewInMemoryStore().(*inMemoryStore),
-		path:          storePath,
+// NewFileStore returns a disk-backed Storer using the provided file path
+func NewFileStore(storagePath string, ilog log.Logger, ttl time.Duration) (Storer, error) {
+	ms := NewInMemoryStore().(*inMemoryStore)
+
+	store := fileStore{
+		path:          storagePath,
+		ilog:          ilog,
+		inMemoryStore: *ms,
 	}
 
-	// Create the external directory for user-generated json
-	storeDir := filepath.Dir(store.path)
-	if _, err := os.Stat(storeDir); err != nil {
-		if err = os.MkdirAll(storeDir, dirFilePerm); err != nil {
-			return nil, fmt.Errorf("store directory in %s could not be created", storeDir)
+	if stat, err := os.Stat(storagePath); err == nil {
+		if now().Sub(stat.ModTime()) > ttl {
+			ilog.Debugf("store file (%s) is older than %v, skipping loading from disk.", storagePath, ttl)
+			return &store, nil
 		}
-	}
 
-	stat, err := os.Stat(store.path)
-	// Store file doesn't exist yet
-	if err != nil {
-		if _, err = os.OpenFile(store.path, os.O_CREATE|os.O_WRONLY, filePerm); err != nil {
-			return nil, fmt.Errorf("store directory not writable: %s", storeDir)
+		store.loadFromDisk()
```

This should return error

---

This could return error in two situations:
- The storage file does not exist. This would happen the first time the integration is executed, and it's not an error that should stop the integration or we should do something else but ignoring it (and log in debug mode).
- The storage file is corrupt or not marshallable. This would happen, e.g., when the integration is updated and the new data model is not compatible with the old one. I think in this case we should also just ignore the error (and log in debug mode).

---

But you should return in case you want to log it on debug mode I guess

---

I think I don't catch you. The `loadFromDisk` already logs in debug ode.

---

Ok, got it. You preferred log everything from the same function. Fixed.